microsoft/testsuites/network/sriov.py

```diff
@@ -310,9 +319,107 @@ def sriov_add_max_nics_validation(
+        vm_nics = self._initialize_nic_info(environment)
+        self._sriov_basic_test(environment, vm_nics)
+
+    @TestCaseMetadata(
+        description="""
+        This case verify VM works well during remove and load sriov modules.
 
-    def _sriov_basic_test(self, environment: Environment) -> None:
+        Steps,
+        1. Provision VM with max network interfaces with enabling accelerated network.
+        2. Do the basic sriov testing.
+        3. Remove sriov module, check network traffic through synthetic nic.
+        4. Load sriov module, check network traffic through VF.
+        """,
+        priority=2,
+        requirement=simple_requirement(
+            min_count=2,
+            min_nic_count=8,
+            network_interface=schema.NetworkInterfaceOptionSettings(
+                data_path=schema.NetworkDataPath.Sriov,
+            ),
+        ),
+    )
+    def sriov_reload_modules_validation(self, environment: Environment) -> None:
+        vm_nics = self._initialize_nic_info(environment)
+        self._sriov_basic_test(environment, vm_nics)
+        module_in_used = ""
+        for node in environment.nodes.list():
+            module_in_used = self._remove_module(node)
+        self._sriov_vf_connection_test(environment, vm_nics, remove_module=True)
+        for node in environment.nodes.list():
+            self._load_module(node, module_in_used)
+        vm_nics = self._initialize_nic_info(environment)
+        self._sriov_vf_connection_test(environment, vm_nics)
+
+    @TestCaseMetadata(
+        description="""
+        This case verify VM works well when provison with max (8) sriov nics.
+
+        Steps,
+        1. Provision VM with max network interfaces with enabling accelerated network.
+        2. Do the basic sriov testing.
+        """,
+        priority=2,
+        requirement=simple_requirement(
+            min_nic_count=8,
+            network_interface=schema.NetworkInterfaceOptionSettings(
+                data_path=schema.NetworkDataPath.Sriov,
+            ),
+        ),
+    )
+    def sriov_disable_enable(self, environment: Environment) -> None:
+        vm_nics = self._initialize_nic_info(environment)
+        self._sriov_basic_test(environment, vm_nics)
+
+    def _initialize_nic_info(
+        self, environment: Environment
+    ) -> Dict[str, Dict[str, NicInfo]]:
+        vm_nics: Dict[str, Dict[str, NicInfo]] = {}
+        for node in environment.nodes.list():
+            node_nic_info = Nics(node)
+            node_nic_info.initialize()
+            for _, node_nic in node_nic_info._nics.items():
+                assert_that(node_nic.lower).described_as(
+                    f"This interface {node_nic.upper} does not have a paired VF."
+                ).is_not_equal_to("")
+            vm_nics[node.name] = node_nic_info._nics
+        return vm_nics
+
+    def _remove_module(self, node: Node) -> str:
+        reload_modules_dict: Dict[str, str] = {
+            "mlx5_core": "mlx5_ib",
```

Comment:
2. firstly we need check the vf use which module, if use mlx5_core, since mlx5_ib depends on it, reload and load mlx5_ib.
3. I try use modprobe --show-depends + recursive, didn't make it. the logic is we need check all the modules which depends on mlx5_core, let 's say module A and B, then check which modules depend on module A and B, till there is no any modules depend on the modules you want to remove, then start to remove, otherwise we will see 'the module is in used' issue, since the mlx modules have simple dependency, so I use dict here.

Reply:
add comments